File: sdk/python/twalk_sdk/disclosure.py
"""The sentence a persona's reply carries to the contact, and how the persona
picks it (ADR 0019, ADR 0031, issue #121).

Every reply a persona drafted reaches the contact with one sentence after
it, on a line of its own, in the language the reply was written in:
*"Rédigé avec mon assistant IA."*, *"Drafted with my AI assistant."*. The
sentence is **not** written by the model — a model can be argued out of
anything — and not composed by the Companion Gateway, which knows no
language and holds no catalogue. The persona **selects** it, because the
persona is the only component that knows what language it wrote in, and it
travels as a field of its own (``data.disclosure``) rather than inside the
body the user may edit. The Gateway appends it at approval.

Three decisions live here rather than in a reader's head.

**The sentences are the contract's**, not this module's:
``contracts/disclosure/v1/sentences.json`` is the authority, and
:data:`SENTENCES` is a copy of it that ``tests/test_disclosure.py`` pins to
the file value for value — the Gateway reads the same file, so the two
components cannot disagree about what a contact is told. A sentence changes
only by a new version directory of the contract, never by an edit here.

**The language is the model's answer, never a library's guess** (ADR 0016:
detection is the model's job). After a persona's handler returns a
suggestion, the SDK asks the same model one more question — the text, and
:data:`LANGUAGE_ASK` as the system prompt — and reads one token back. An
author who *knows* the language (a persona that only ever writes French)
says so on the :class:`~twalk_sdk.envelope.Suggestion` and skips the ask.

**A language with no sentence is no suggestion at all.** ``sentence_for``
answers ``None``, and the persona loop raises :class:`DisclosureError`,
which is **not transient**: the model would answer the same language to the
same reply, so the trigger is terminated on the first delivery with one
``ERROR`` line naming the tag, exactly as a model that spent its whole
budget reasoning is (:mod:`twalk_sdk.completion`). Falling back to English,
or to the user's own language, was rejected in ADR 0031: a disclosure the
contact cannot read is a sentence nobody reads, and a named refusal is a
contribution request an operator can act on with one line.

**The ask has a budget, and a reasoning model needs it raised.** The ask
carries :data:`LANGUAGE_ASK_MAX_TOKENS` — five, because one tag is all it
needs back — unless the operator's provider parameters name ``max_tokens``,
which are merged last and then govern the ask exactly as they govern the
reply. A reasoning model spends its first tokens thinking, so with five it
answers nothing and :mod:`twalk_sdk.completion` reports it as having spent
its whole budget reasoning — #162's failure again, one call later and after
the reply was already drafted and billed. The loop wraps that in
:class:`LanguageAskFailed` so the ``ERROR`` line says which of the two calls
it was and names the remedy: ``TWALK_LLM_PARAMS={"max_tokens": N}`` (from
``HERMES_LLM_PARAMS`` in a deployment), which is documented as required on
a reasoning model rather than left for an operator to infer from a log.
"""

from __future__ import annotations

from typing import Dict, Optional

from .completion import BUDGET_REMEDY, LlmError, LlmSpentItsBudgetThinking

#: The contract's five sentences, keyed by the primary language subtag. A
#: copy of ``contracts/disclosure/v1/sentences.json``, pinned to it by
#: ``tests/test_disclosure.py``: if the two disagree, the contract wins and
#: the test says so.
SENTENCES: Dict[str, str] = {
    "en": "Drafted with my AI assistant.",
    "fr": "Rédigé avec mon assistant IA.",
    "it": "Scritto con il mio assistente IA.",
    "es": "Redactado con mi asistente de IA.",
    "de": "Verfasst mit meinem KI-Assistenten.",
}

#: The schema's cap on ``data.disclosure``. Named here because
#: :mod:`twalk_sdk.envelope` subtracts it, plus the newline, from the body's
#: own cap, so that body + line never exceeds ``final.body``'s 65 536 when
#: the Companion Gateway appends the sentence at approval.
DISCLOSURE_MAX_CHARS = 200

#: What the persona asks its model after the reply is drafted, verbatim: one
#: system prompt, the reply as the user message, one token back. The
#: wording is load-bearing beyond this package — the test harness's stub
#: model recognises the ask by the phrase "exactly one language tag" — so it
#: is a constant and not a sentence composed at the call site.
LANGUAGE_ASK = "Answer with exactly one language tag among en, fr, it, es, de, or other."

#: The ``max_tokens`` the ask carries when the operator's provider
#: parameters name none: one tag is all it needs back. Provider parameters
#: are merged last (:mod:`twalk_sdk.config`), so ``TWALK_LLM_PARAMS``
#: naming ``max_tokens`` replaces this for the ask as it does for the reply
#: — which is what a reasoning model needs, since it thinks before it
#: answers and five tokens of thinking is no answer.
LANGUAGE_ASK_MAX_TOKENS = 5

#: The one answer that is not a tag: the model saying the reply is in none
#: of the five. Read as "no language", so the refusal names what the model
#: said rather than inventing a tag it did not.
OTHER = "other"

#: How much of the model's first token a refusal echoes. The token is what
#: the model answered to a one-word question, and a model that ignored the
#: system prompt and echoed or translated the draft puts the first word of
#: the persona's reply there — not contact text, but message content, so
#: the log line carries at most this much of it.
TAG_ECHO_MAX_CHARS = 32

#: What is trimmed off the model's first token before it is read as a tag:
#: a model asked for one word still likes to end it with a period or wrap
#: it in quotes or backticks.
_PUNCTUATION = ".,;:!?\"'`*()[]{}<>"


class DisclosureError(RuntimeError):
    """The reply is in a language the contract has no sentence for, so there
    is no suggestion.

    **Not transient**, and the persona loop reads exactly that: the model
    answered which language it wrote in, and asking it again about the same
    reply returns the same language. One ``ERROR`` line, the delivery
    terminated, never a retry — the shape #162 set for a model that spent
    its whole budget reasoning, because here too every attempt is billed.
    """

    transient = False

    def __init__(self, tag: Optional[str], *, declared_by_persona: bool = False) -> None:
        self.tag = tag
        who = "the persona declared" if declared_by_persona else "the model answered"
        super().__init__(
            f"disclosure has no sentence for the language {who}: {_echo(tag)}. "
            f"A reply the contact cannot be told was drafted with an AI assistant "
            f"is no suggestion at all (ADR 0031); the languages with a sentence "
            f"are {', '.join(SENTENCES)}, from contracts/disclosure/v1/sentences.json"
        )


class LanguageAskFailed(LlmError):
    """The language ask — not the reply — is the completion that failed.

    The ask is a second call to the same model, made after the reply was
    drafted, and it fails the same four ways the reply can
    (:mod:`twalk_sdk.completion`). Reported bare, its ``ERROR`` line reads
    exactly like the reply's own failure, and an operator on a reasoning
    model — where the ask's five-token budget is spent thinking before a tag
    is written, every time — would look for a reply that was in fact drafted
    fine. So the loop wraps the cause: the line says it was the ask, says
    the reply was drafted, and names the remedy, which is the same
    ``max_tokens`` the reply's own budget lives in.

    Whether it is retried is the cause's decision, not this wrapper's:
    :attr:`transient` is copied from it, so a model that spent its budget
    thinking about a one-word question terminates the delivery the way it
    does about the reply, and an endpoint that was briefly away is retried.
    """

    def __init__(self, cause: LlmError) -> None:
        self.cause = cause
        self.transient = cause.transient
        if isinstance(cause, LlmSpentItsBudgetThinking):
            what = "exhausted the model's budget"
        else:
            what = "did not get an answer"
        super().__init__(
            f"the language ask {what}: the one-word question the SDK puts to the "
            f"model after the reply is drafted, to select the disclosure's "
            f"language (ADR 0031) — the reply itself was drafted and is not the "
            f"problem. The ask carries {LANGUAGE_ASK_MAX_TOKENS} tokens unless "
            f"TWALK_LLM_PARAMS names max_tokens, which then governs the ask as "
            f"it governs the reply; a reasoning model spends its first tokens "
            f"thinking, so on one that budget must be set — {BUDGET_REMEDY}. "
            f"The ask failed with: {cause}"
        )


def _echo(tag: Optional[str]) -> str:
    """What a refusal names as the language: the tag, at most
    :data:`TAG_ECHO_MAX_CHARS` of it, and ``other`` for none."""
    if not tag:
        return OTHER
    if len(tag) <= TAG_ECHO_MAX_CHARS:
        return tag
    return tag[:TAG_ECHO_MAX_CHARS] + "…"


def primary_subtag(tag: str) -> str:
    """The language part of a BCP 47 tag: ``fr-CA`` and ``fr_CA`` are
    ``fr``; ``FR`` is ``fr``."""
    return tag.strip().lower().replace("_", "-").split("-", 1)[0]


def sentence_for(tag: Optional[str]) -> Optional[str]:
    """The sentence for a language, or ``None`` when the contract has none.

    Keyed on the primary subtag, so a model or an author saying ``fr-CA``
    gets the French sentence: the sentence discloses a fact, and the fact is
    the same in Québec. ``None`` in, ``None`` out — a model that answered
    ``other`` declared no language.
    """
    if not tag:
        return None
    return SENTENCES.get(primary_subtag(tag))


def parse_language_answer(answer: str) -> Optional[str]:
    """The tag in a model's answer to :data:`LANGUAGE_ASK`, or ``None``.

    Lower-cased, trimmed, the first whitespace-separated token with its
    surrounding punctuation removed: ``" FR\\n"`` is ``fr``, ``"fr."`` is
    ``fr``, ``"fr-CA"`` is ``fr-ca`` (mapped to the sentence by
    :func:`sentence_for`). ``other`` and an empty answer are ``None``.
    Anything else is returned as the model said it — ``ja``, or a word that
    is not a tag at all — so that the refusal it leads to names the answer
    rather than a guess about it.
    """
    tokens = answer.strip().lower().split()
    if not tokens:
        return None
    token = tokens[0].strip(_PUNCTUATION)
    if not token or token == OTHER:
        return None
    return token

File: companion-gateway/src/disclosure.rs
//! The disclosure: the one sentence every reply a persona drafted carries to
//! the contact, and what this Gateway does with it (ticket #121, ADR 0019,
//! ADR 0031).
//!
//! # What the Gateway does, and what it deliberately does not
//!
//! ADR 0031 closed three of the four homes the sentence could have had, and
//! one of the three was **the Gateway composing it**: no event carries a
//! language, no component detects one, and there is no server-side message
//! catalogue in this repository for a legally significant text. So the
//! sentence is *selected* by whoever knows the language of the reply — the
//! persona, on the SDK path; Hermes's own answer, on the return path of ADR
//! 0032 — and travels as `data.disclosure` on `persona.suggest.produced`, a
//! field of its own and never inside the body. What this Gateway does is
//! **append** it: at approval, `final.body` becomes the body, a newline and
//! the sentence, and the approved event carries the sentence as
//! `data.disclosure` too, so the bus records what the contact received and
//! what part of it was the disclosure.
//!
//! # Whose sentences these are
//!
//! `contracts/disclosure/v1/sentences.json` — the contract's, not this
//! crate's. The file is compiled in with `include_str!` for the reason the
//! connection kinds are: one authority, two readers (the Python SDK is the
//! other), and if a component and the contract disagree, the contract wins.
//! A sentence changes only through a new version directory, never an in-place
//! edit; the tests below pin the five keys to the five languages the
//! Companion ships, because a sixth sentence with no interface behind it, or
//! a sixth interface language with no sentence, is a gap somebody should
//! have to argue for.
//!
//! The Gateway reads the table for one thing only: **Hermes's answer names a
//! language and not a sentence**, because Hermes is outside this deployment
//! and holds no copy of the contract. [`sentence_for`] maps that tag to its
//! primary subtag (`fr-CA` is French) and looks the sentence up; a language
//! outside the five is a named, counted refusal and no suggestion at all,
//! which is ADR 0031's posture on the SDK path carried over: a suggestion
//! that cannot be disclosed is one that should not exist.
//!
//! # The switch
//!
//! Removable **only globally**, never from a single message, and every
//! removal a dated, attributed row in an append-only journal of its own
//! ([`crate::store`]'s `disclosure_decision`): a sibling of the consent
//! journal, sharing its triggers and its `occurred_at`/`actor`/`reason`
//! shape, and deliberately neither the consent journal itself nor the
//! settings table, whose single-row upsert forgets who decided and when.
//! No row means **on**: the disclosure ships on by default and the journal
//! records the decisions to turn it off and back.

use std::collections::BTreeMap;
use std::sync::OnceLock;

use serde_json::Value;

/// The contract's own table: `contracts/disclosure/v1/sentences.json`, one
/// sentence per language tag, verbatim.
pub const SENTENCES_JSON: &str = include_str!("../../contracts/disclosure/v1/sentences.json");

/// The contract's cap on `data.disclosure` (`maxLength: 200` on both
/// schemas). What a reply's body cap is reduced by, plus the newline, so the
/// appended `final.body` never exceeds the contract's 65 536.
pub const MAX_CHARS: usize = 200;

/// The longest a `reason` on a switch decision may be — the consent
/// journal's own cap on the same member.
pub const MAX_REASON_CHARS: usize = 1024;

fn sentences() -> &'static BTreeMap<String, String> {
    static SENTENCES: OnceLock<BTreeMap<String, String>> = OnceLock::new();
    SENTENCES.get_or_init(|| {
        serde_json::from_str(SENTENCES_JSON)
            .expect("contracts/disclosure/v1/sentences.json is a JSON object of strings")
    })
}

/// The language tags the contract holds a sentence for, in the file's own
/// (sorted) order.
pub fn languages() -> Vec<&'static str> {
    sentences().keys().map(String::as_str).collect()
}

/// The primary subtag of a BCP 47 language tag: `fr-CA` is `fr`, `sr-Latn-RS`
/// is `sr`. The region and the script say nothing about which sentence a
/// contact can read.
pub fn primary_subtag(tag: &str) -> &str {
    tag.split('-').next().unwrap_or(tag).trim()
}

/// The sentence for a language tag, by its primary subtag, or `None` when the
/// contract holds none — which the caller turns into a refusal, never into a
/// sentence in another language (ADR 0031).
pub fn sentence_for(tag: &str) -> Option<&'static str> {
    sentences()
        .get(&primary_subtag(tag).to_ascii_lowercase())
        .map(String::as_str)
}

/// Whether a sentence is one of the contract's five, verbatim.
///
/// The check the SDK path needs and the Hermes path does not: on the
/// Hermes path the sentence is *selected here* from a language tag, so it
/// cannot be anything else, but on the SDK path it arrives on the bus as
/// `data.disclosure`, and the schema says only `string, 1..200`. The
/// reference bus has no authentication, so "what a contact is told is one
/// of the contract's own sentences" — the property ADR 0031 exists for, and
/// the reason the model never writes the sentence — holds at approval only
/// if this Gateway asks. Verbatim, not trimmed, not case-folded: a sentence
/// that differs from the contract's by a character is not the contract's.
pub fn is_contract_sentence(candidate: &str) -> bool {
    sentences().values().any(|sentence| sentence == candidate)
}

/// The body as the contact receives it: the reply, then the sentence on a
/// line of its own. After and not before, because a prefix is precisely what
/// a bridge's relay mode does and ADR 0025 rejected that by name.
pub fn append(body: &str, sentence: &str) -> String {
    format!("{body}\n{sentence}")
}

/// The switch as the journal answers it: the last decision, or the default.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DisclosureState {
    pub enabled: bool,
    /// When the last decision was taken, RFC 3339; `None` when none ever
    /// was, which is the default and says "on since the deployment began".
    pub since: Option<String>,
    /// Who took it: the deployment's owner, as a consent decision's `actor`.
    pub actor: Option<String>,
    pub reason: Option<String>,
}

impl DisclosureState {
    /// The state a store with no decision in it holds: on, and nobody
    /// decided.
    pub const DEFAULT: Self = Self {
        enabled: true,
        since: None,
        actor: None,
        reason: None,
    };

    /// The journal's own vocabulary for a state: what `new_state` holds.
    /// The one place the two words are spelled, so the writer and the reader
    /// in [`crate::store`] cannot drift from each other.
    pub fn word(enabled: bool) -> &'static str {
        if enabled {
            "on"
        } else {
            "off"
        }
    }

    /// The reverse: a `new_state` read back, or `None` for a word the
    /// journal never writes.
    pub fn enabled_from(word: &str) -> Option<bool> {
        match word {
            "on" => Some(true),
            "off" => Some(false),
            _ => None,
        }
    }
}

/// One decision as a client states it: `{"enabled": bool, "reason"?: string}`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DisclosureUpdate {
    pub enabled: bool,
    pub reason: Option<String>,
}

/// What is wrong with a switch request. One code, `malformed_request`, and a
/// sentence naming the member, because every one of these is something the
/// caller can fix.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Invalid(pub String);

impl Invalid {
    pub fn code(&self) -> &'static str {
        "malformed_request"
    }

    pub fn message(&self) -> &str {
        &self.0
    }
}

/// Parses a switch request. A closed object: `enabled` is required and a
/// boolean, `reason` is optional and at most [`MAX_REASON_CHARS`], and any
/// other member is refused rather than ignored — a client that sent
/// `{"disclosure": false}` should learn that nothing happened.
pub fn parse_update(body: &Value) -> Result<DisclosureUpdate, Invalid> {
    let object = body.as_object().ok_or_else(|| {
        Invalid("the request body is not a JSON object: send {\"enabled\": true|false}".to_owned())
    })?;
    for member in object.keys() {
        if !matches!(member.as_str(), "enabled" | "reason") {
            return Err(Invalid(format!(
                "the request body has the unknown member {member:?}: a disclosure decision \
                 carries enabled and an optional reason"
            )));
        }
    }
    let enabled = match object.get("enabled") {
        Some(Value::Bool(enabled)) => *enabled,
        Some(_) => {
            return Err(Invalid(
                "enabled is a boolean: true appends the disclosure to every approved reply, \
                 false stops it for every reply until it is turned on again"
                    .to_owned(),
            ))
        }
        None => return Err(Invalid("enabled is required".to_owned())),
    };
    let reason = match object.get("reason") {
        None | Some(Value::Null) => None,
        Some(Value::String(reason)) => {
            let count = reason.chars().count();
            if count > MAX_REASON_CHARS {
                return Err(Invalid(format!(
                    "reason is {count} characters, and the limit is {MAX_REASON_CHARS}"
                )));
            }
            Some(reason.clone())
        }
        Some(_) => return Err(Invalid("reason is a string when it is present".to_owned())),
    };
    Ok(DisclosureUpdate { enabled, reason })
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn the_contracts_languages_are_exactly_the_five_the_companion_ships() {
        // Two readers of one table, and a third list in `settings.rs`: this
        // is what keeps them one. A sixth sentence with no interface behind
        // it, or a sixth interface language with no sentence, fails here
        // rather than being discovered by a user.
        let mut companion: Vec<&str> = crate::settings::LANGUAGES
            .iter()
            .map(|language| language.as_str())
            .collect();
        companion.sort_unstable();
        assert_eq!(languages(), companion);
    }

    #[test]
    fn every_sentence_is_one_plain_line_the_contract_allows() {
        for (tag, sentence) in sentences() {
            assert!(!sentence.is_empty(), "{tag}: an empty disclosure");
            assert!(
                sentence.chars().count() <= MAX_CHARS,
                "{tag}: the sentence is longer than the contract's {MAX_CHARS}"
            );
            assert!(
                !sentence.contains('\n'),
                "{tag}: the disclosure is one line, on a line of its own"
            );
            assert_eq!(
                sentence.trim(),
                sentence,
                "{tag}: no leading or trailing blank"
            );
            assert!(
                !sentence.to_lowercase().contains("twalk"),
                "{tag}: the sentence describes the message and not the user's infrastructure \
                 (ADR 0031): {sentence}"
            );
        }
    }

    #[test]
    fn the_sentences_are_the_contracts_own() {
        // The five, verbatim, so that a change to the contract file is a
        // change a reviewer sees here as well.
        assert_eq!(sentence_for("en"), Some("Drafted with my AI assistant."));
        assert_eq!(sentence_for("fr"), Some("Rédigé avec mon assistant IA."));
        assert_eq!(
            sentence_for("it"),
            Some("Scritto con il mio assistente IA.")
        );
        assert_eq!(
            sentence_for("es"),
            Some("Redactado con mi asistente de IA.")
        );
        assert_eq!(
            sentence_for("de"),
            Some("Verfasst mit meinem KI-Assistenten.")
        );
    }

    #[test]
    fn a_refined_tag_is_read_by_its_primary_subtag() {
        assert_eq!(sentence_for("fr-CA"), sentence_for("fr"));
        assert_eq!(sentence_for("de-CH"), sentence_for("de"));
        assert_eq!(sentence_for("sr-Latn-RS"), None, "no Serbian sentence");
        assert_eq!(primary_subtag("pt-BR"), "pt");
        assert_eq!(primary_subtag("en"), "en");
    }

    #[test]
    fn a_language_outside_the_five_has_no_sentence_and_is_not_defaulted() {
        for tag in ["ja", "pt", "pt-BR", "ar", "zh", "", "français"] {
            assert_eq!(sentence_for(tag), None, "{tag:?}");
        }
    }

    #[test]
    fn only_the_contracts_own_sentences_verbatim_are_sentences() {
        for (_, sentence) in sentences() {
            assert!(is_contract_sentence(sentence), "{sentence}");
        }
        for forged in [
            // A persona's own wording, the length of a real sentence.
            "Written by an assistant you can trust.",
            // One character off the contract's: a period lost, a case
            // changed, a blank added. Not the contract's.
            "Rédigé avec mon assistant IA",
            "rédigé avec mon assistant IA.",
            " Rédigé avec mon assistant IA.",
            "Rédigé avec mon assistant IA.\n",
            // The schema refuses an empty member and so does this.
            "",
            // A language tag is what Hermes sends and not what a persona
            // sends: a tag on the bus is not a sentence.
            "fr",
        ] {
            assert!(!is_contract_sentence(forged), "{forged:?}");
        }
    }

    #[test]
    fn the_sentence_goes_after_the_body_on_a_line_of_its_own() {
        assert_eq!(
            append(
                "Pas de problème, à 20h ! 👍",
                "Rédigé avec mon assistant IA."
            ),
            "Pas de problème, à 20h ! 👍\nRédigé avec mon assistant IA."
        );
    }

    #[test]
    fn a_decision_is_a_boolean_and_an_optional_reason() {
        assert_eq!(
            parse_update(&json!({ "enabled": false, "reason": "a test" })),
            Ok(DisclosureUpdate {
                enabled: false,
                reason: Some("a test".to_owned())
            })
        );
        assert_eq!(
            parse_update(&json!({ "enabled": true, "reason": null })),
            Ok(DisclosureUpdate {
                enabled: true,
                reason: None
            })
        );
        for (body, names) in [
            (json!([]), "JSON object"),
            (json!({}), "enabled is required"),
            (json!({ "enabled": "off" }), "enabled is a boolean"),
            (
                json!({ "enabled": false, "reason": 3 }),
                "reason is a string",
            ),
            (
                json!({ "enabled": false, "reason": "x".repeat(1025) }),
                "1024",
            ),
            (json!({ "disclosure": false }), "unknown member"),
        ] {
            let refusal = parse_update(&body).expect_err(&body.to_string());
            assert_eq!(refusal.code(), "malformed_request");
            assert!(
                refusal.message().contains(names),
                "{body}: {}",
                refusal.message()
            );
        }
    }

    #[test]
    fn the_default_is_on_and_nobody_decided() {
        let state = DisclosureState::DEFAULT;
        assert_eq!(
            state,
            DisclosureState {
                enabled: true,
                since: None,
                actor: None,
                reason: None
            }
        );
        assert_eq!(DisclosureState::word(state.enabled), "on");
        assert_eq!(DisclosureState::word(false), "off");
        assert_eq!(DisclosureState::enabled_from("on"), Some(true));
        assert_eq!(DisclosureState::enabled_from("off"), Some(false));
        assert_eq!(DisclosureState::enabled_from("ON"), None);
    }
}
